nexus/src/app/update/sp_updater.rs

```diff
@@ -0,0 +1,345 @@
+                        let progress = if bytes_received > total_bytes {
+                            warn!(
+                                self.log, "nonsense SP progress";
+                                "bytes_received" => bytes_received,
+                                "total_bytes" => total_bytes,
+                            );
+                            None
+                        } else if total_bytes == 0 {
+                            None
+                        } else {
+                            Some(
+                                f64::from(bytes_received)
+                                    / f64::from(total_bytes),
+                            )
+                        };
+                        self.progress.send_replace(Some(
+                            UpdateProgress::InProgress { progress },
+                        ));
+                        tokio::time::sleep(STATUS_POLL_INTERVAL).await;
+                        continue;
+                    } else {
+                        format!("different update is now in progress ({id})")
+                    }
+                }
+                SpUpdateStatus::Complete { id } => {
+                    if id == self.update_id {
+                        self.progress.send_replace(Some(
+                            UpdateProgress::InProgress { progress: Some(1.0) },
+                        ));
+                        return Ok(());
+                    } else {
+                        format!("different update is now in complete ({id})")
+                    }
+                }
+                SpUpdateStatus::None => {
+                    "update status lost (did the SP reset?)".to_string()
+                }
+                SpUpdateStatus::Aborted { id } => {
+                    if id == self.update_id {
+                        "update was aborted".to_string()
+                    } else {
+                        format!("different update is now in complete ({id})")
+                    }
+                }
+                SpUpdateStatus::Failed { code, id } => {
+                    if id == self.update_id {
+                        format!("update failed (error code {code})")
+                    } else {
+                        format!("different update failed ({id})")
+                    }
+                }
+                SpUpdateStatus::RotError { id, message } => {
+                    if id == self.update_id {
+                        format!("update failed (rot error: {message})")
+                    } else {
+                        format!("different update failed with rot error ({id})")
+                    }
+                }
+            };
+
+            self.progress.send_replace(Some(UpdateProgress::Failed(
+                error_message.clone(),
+            )));
+            return Err(SpUpdateError::FailedToComplete(error_message));
+        }
+    }
+
+    async fn finalize_update_via_reset_one_mgs(
```

Comment:
What do you think about waiting for it to come back up and verifying that it reports that it's running the new version?  I'm just thinking that if this ever _didn't_ happen, we'd want to know about it promptly as part of the update process.  Also, it'd be nice to know that after a successful SP update, the SP is running.  If we don't wait for that I could imagine the next step in an update process failing spuriously with "something is wrong -- the SP is not up".

Reply:
That's a good idea; I'll add that. FWIW the `reset` operation already waits until the SP comes back and is able to answer messages, so we'd already fail if the SP didn't come up at all. But without the extra check we could still be running the old version (at least in theory, although I don't think that's possible in practice today given how the SP boots).

Reply:
We chatted about this offline and opted _not_ to do this: `SpUpdater` doesn't currently know _what_ version it's updating the SP to, so it would either need to open the SP archive with `hubtools` and inspect, or accept the version from the caller (and trust that the caller told it a version that matches the SP archive). We don't yet have the mechanism by which `SpUpdater` gets its data, but it will almost certainly require using `hubtools` to inspect the archive _before_ we get here, and we don't really want to repeat that inspection. The next inventory collection taken after any SP update will report the version it's running, and `SpUpdater`'s caller should be in a position to trigger such a collection.